src/Grid.ts

```diff
@@ -0,0 +1,110 @@
+import { Subscription } from '@dojo/shim/Observable';
+import { v, w } from '@dojo/widget-core/d';
+import { DNode, PropertyChangeRecord } from '@dojo/widget-core/interfaces';
+import { RegistryMixin }  from '@dojo/widget-core/mixins/Registry';
+import { theme, ThemeableMixin, ThemeableProperties } from '@dojo/widget-core/mixins/Themeable';
+import WidgetBase, { diffProperty } from '@dojo/widget-core/WidgetBase';
+import WidgetRegistry from '@dojo/widget-core/WidgetRegistry';
+import DataProviderBase, { Options } from './bases/DataProviderBase';
+import Body from './Body';
+import Cell from './Cell';
+import Header from './Header';
+import HeaderCell from './HeaderCell';
+import { DataProperties, HasColumns } from './interfaces';
+import Row from './Row';
+
+import * as css from './styles/grid.m.css';
+
+export const GridBase = ThemeableMixin(RegistryMixin(WidgetBase));
+
+/**
+ * @type GridProperties
+ *
+ * Properties that can be set on a Grid
+ *
+ * @property columns		Column definitions
+ * @property dataProvider	An observable object that responds to events and returns {@link DataProperties}
+ */
+export interface GridProperties extends ThemeableProperties, HasColumns {
+	registry?: WidgetRegistry;
+	dataProvider: DataProviderBase<any, Options>;
+}
+
+const gridRegistry = new WidgetRegistry();
+gridRegistry.define('header', Header);
+gridRegistry.define('header-cell', HeaderCell);
+gridRegistry.define('body', Body);
+gridRegistry.define('row', Row);
+gridRegistry.define('cell', Cell);
+
+@theme(css)
+class Grid extends GridBase<GridProperties> {
+	private _data: DataProperties<any>;
+	private _subscription: Subscription;
+
+	constructor() {
+		super();
+
+		this.registries.add(gridRegistry);
+	}
+
+	@diffProperty('dataProvider')
+	protected diffPropertyDataProvider(previousValue: DataProviderBase<any, Options>, value: DataProviderBase<any, Options>): PropertyChangeRecord {
```

Comment:
You should only need to observe on the provider if it has changed since the last render. At the moment every `render` this will be calling `unsubscribe` and observing on the same provider. 

In addition at the moment this diff will also cause an infinite loop of rendering as the subscriber is calling `invalidate`.